### ISSUE TITLE
math builtins

### ISSUE THREAD

**Issue by ewmailing** [pallene-lang/pallene#490]
### Discussed in https://github.com/pallene-lang/pallene/discussions/489

<div type='discussions-op-text'>

<sup>Originally posted by **ewmailing** October  4, 2021</sup>
Hi, I was trying to implement some code for my little experiment project that depend on math library functions, specifically: exp, log, pow.
I figured out how to hack these into Pallene using sqrt as the template. It's not a lot of code, but would you like me to submit a pull request?

I should note that my math.log is not fully correct because I could only figure out how to support one parameter (mapping to C's log.) But I figure at the moment, this incomplete implementation is still no worse than no implementation. For two parameters, I haven't figured out:
- How to tell the system that a second optional float parameter is supported (and how to refer to it)
- How to dynamically vary the output depending if the second parameter is 1, 2, 10, or something else (so the generated code is one of: log(x), log2(x), log10(x), or log(x)/log(base)

Similarly, math.abs also needs to handle floats and integers differently (call fabs or abs (or is it labs)). 

Thanks</div>

**Comment by hugomg** [comment]
>  I needed a few more math functions, so I tried implementing a few more (ceil, floor, fmod, modf).
I got stuck on modf.

More functions would be nice :) However, I think  it would be easier to review if we put those on separate pull request(s).

> I'm getting an error that looks like:

I don't know what is the problem but usually a Lua stack trace like that points to a bug in the compiler itself. The sort of type errors we can get when programming in a dynamic language... If it is something simple like a typo then luacheck might catch it.

> Also, does the infrastructure exist to export the constants like math.pi, math.huge, etc.?

After a brief look, my guess is that we may not need to change the typechecker (checker.lua) but we may need to do some small changes to builtins.lua and to_ir.lua.

The builtins.lua has a table called `builtins.functions`, which sounds like it accepts only functions. However, I'm not sure if we actually require them to be functions. Maybe this could just as well have been called `builtins.values`? I'm not 100% sure.

In the to_ir.lua, one part that we will have to deal with is the ast.Exp.Var case in exp_to_value. Currently, returning builtins as a value is not yet implemented (#337). However, we could check if the type of the builtin is a function, or if it is a number. That way we can implement constants like `e` and `pi` while still leaving the builtin functions for later.

**Comment by hugomg** [comment]
> Okay,, I figured out the first bug. It was my mistake. I missed a pair of parenthesis after T.Integer in builtins.lua.

I'll see this as an incentive towards converting the compiler to Pallene. Who knows, maybe one day we'll do that... :)

> But I'm still waiting on what I should do for the original abs, exp, low, pow PR. Have you had a chance to look at my latest commits?

Oh, I hadn't noticed that this was ready for review. Looking at it now, it seems that the tests are still broken. I think it might be the tests for the Pallene->Lua translator because of that math.ln thing we did? Could you please look into that? 

Once you're done, feel free to notify me using Github's "request review" feature.

![Captura de tela de 2021-10-13 00-18-36](https://user-images.githubusercontent.com/1452192/137061557-b49823b8-b8ec-42c8-a6c6-d63e45f8c22d.png)
